pyk/src/pyk/kast/inner.py

```diff
@@ -877,20 +877,21 @@ def _var_occurence(_term: KInner) -> None:
     return _var_occurrences
 
 
-# TODO replace by method that does not reconstruct the AST
 def collect(callback: Callable[[KInner], None], kinner: KInner) -> None:
     """Collect information about a given term traversing it bottom-up using a function with side effects.
 
     Args:
         callback: Function with the side effect of collecting desired information at each AST node.
         kinner: The term to traverse.
     """
+    subterms = [kinner]
 
-    def f(kinner: KInner) -> KInner:
-        callback(kinner)
-        return kinner
+    for term in subterms:
+        subterms.extend(term.terms)  # noqa: B909
 
-    bottom_up(f, kinner)
+    while subterms:
+        term = subterms.pop()
+        callback(term)
```

Comment:
Could this be done in one loop instead of two?

```
for term in subterms:
    subterms.extend(term.terms)
    callback(term)
```

Reply:
It could in a top-down manner, but I don't think it's possible bottom-up.

Reply:
I pushed a commit, please take a look: 53a8cc2

Reply:
Uses downstream (checked off means "Looks ok for top-down"):

- [X] https://github.com/runtimeverification/kontrol/blob/cc9afb725355053938e9bd788860f0bc4e7516e0/src/kontrol/foundry.py#L972
- [X] no direct uses in KEVM
- [X] no direct uses in KWasm
- [X] no direct uses in mx-semantics
- [X] no direct uses in mir-semantics
- [X] no direct uses in Komet
- [X] no direct uses in KRiscV

Reply:
Good to resolve.